sky/provision/azure/instance.py

```diff
@@ -0,0 +1,215 @@
+        resource_group_name=provider_config['resource_group'])
+
+    nodes = [
+        _extract_metadata(vm, provider_config)
+        for vm in filter(match_tags, vms)
+    ]
+    return {node['name']: node for node in nodes}
+
+
+def stop_instances(
+    cluster_name: str,
+    provider_config: Optional[Dict[str, Any]] = None,
+    worker_only: bool = False,
+) -> None:
+    assert provider_config is not None, cluster_name
+    subscription_id = provider_config['subscription_id']
+    resource_group = provider_config['resource_group']
+
+    instances_to_stop = _filter_instances(
+        cluster_name,
+        provider_config,
+        {},
+        worker_only,
+    )
+    compute_client = azure.get_client('compute', subscription_id)
+
+    stop = get_azure_sdk_function(
+        client=compute_client.virtual_machines,
+        function_name='deallocate',
+    )
+    for node_id in instances_to_stop.keys():
+        stop(resource_group_name=resource_group, vm_name=node_id)
+
+
+def terminate_instances(
+    cluster_name: str,
+    provider_config: Optional[Dict[str, Any]] = None,
+    worker_only: bool = False,
+) -> None:
+    """See sky/provision/__init__.py"""
+    assert provider_config is not None, cluster_name
+    subscription_id = provider_config['subscription_id']
+    resource_group = provider_config['resource_group']
+    compute_client = azure.get_client('compute', subscription_id)
+    network_client = azure.get_client('network', subscription_id)
+
+    if not worker_only:
+        # If we are going to terminate all instances, then
+        # we just take down the resource group in case of resource leak.
+        resource_client = azure.get_client('resource', subscription_id)
+        resource_client.resource_groups.begin_delete(resource_group)
+        return
+
+    instances_to_terminate = _filter_instances(
+        cluster_name,
+        provider_config,
+        {},
+        worker_only,
+    )
+
+    disks: Set[str] = set()
+
+    for node_id in instances_to_terminate:
+        vm = compute_client.virtual_machines.get(
+            resource_group_name=resource_group, vm_name=node_id)
+        disks.update(d.name for d in vm.storage_profile.data_disks)
+        disks.add(vm.storage_profile.os_disk.name)
+
+    # delete machine, must wait for this to complete
+    delete = get_azure_sdk_function(client=compute_client.virtual_machines,
+                                    function_name='delete')
+
+    operations = []
+    for node_id in instances_to_terminate:
+        try:
+            operations.append(
+                delete(resource_group_name=resource_group, vm_name=node_id))
+        except Exception as e:  # pylint: disable=broad-except
+            logger.warning('Failed to delete VM: {}'.format(e))
+    for opr in operations:
+        try:
+            opr.wait()
+        except Exception as e:  # pylint: disable=broad-except
+            logger.warning('Failed to delete VM: {}'.format(e))
+
```

Comment:
If we only terminate the VM, and guarantee the `worker_only=False` (i.e., directly deleting the resource group) will be called right after this function finish, will there be any leakage?

If that will not cause leakge, maybe we can change the API from `worker_only` to `worker_first`, i.e., if the `worker_first=True` we first terminate the workers and then terminate the whole cluster in the same function.